app/utils/entity_utils.rb

```diff
@@ -184,26 +197,64 @@ def parse_specs(specs)
   def validate(validators, val, field)
     validators.reduce([]) do |res, validator|
       err = validator.call(val, field)
-      res.push(err) unless err.nil?
+      res.push({field: field.to_s, code: err[:code], msg: err[:msg]}) unless err.nil?
       res
     end
   end
 
+  def validate_all(fields, input)
+    fields.reduce([]) do |errs, (name, spec)|
+      errors = validate(spec[:validators], input[name], name).map { |err|
+        err[:field] = "#{name.to_s}"
+        err
+      }
+
+      nested_errors =
+        if spec[:collection].present?
+          input[name].each_with_index.flat_map { |v, i|
```

Comment:
Can you come up with a way to use reduce and mutations instead of flat_map and map that create temporary collections in the way? This code gets run a lot so one of the design goals of the original version was to favor one pass processing and mutations where possible. This is of course much more complex scenario to handle.